### ISSUE TITLE
[Select] Allow positioning of options beneath trigger

### ISSUE THREAD

**Issue by afridley** [radix-ui/primitives#1247]
### Overview
Currently the Select menu will position the currently selected option in the list on top of the trigger. This means that depending on the selected option the menu changes positions.

There are times where designs from clients who would like to see the menu positioned below the trigger cannot be currently achieved.

I would like to see an option that will place the menu just below the trigger as seen in the examples below.

### Examples in other libraries
https://mui.com/components/selects/
https://react-select.com/home

Notice the menu appears beneath the trigger

### Who does this impact? Who is this for?
This would be an option for all users.
Anyone who has designs or wishes the select menu to appear below the trigger.

**Comment by ArniyhJS** [comment]
I wish this could be looked into.
thanks

**Comment by KenEucker** [comment]
I think a solution, or hack, for this might be to reorder the list putting the selected item at the top so that the dropdown always opens from the bottom of the input. I am going to try this and will report back with a snippet if it works out.

**Comment by benoitgrelard** [comment]
Hey @afridley,

Thanks for suggesting this.
It was very much a design decision to prioritise the UX as it currently stands.
However, Radix should be as customisable as possible, so we should look into seeing if we can provide the ability to do what you are asking.

**Comment by benoitgrelard** [comment]
> @benoitgrelard I appreciate all the work! I'm wondering if you are able to give an estimate on when this could be supported?

Not at the moment I'm afraid.

**Comment by Johnhersh** [comment]
I definitely need this too. Using a dropdown instead of this right now to achieve the desired effect. I feel the way it's currently implemented is too opinionated

**Comment by UTkzhang** [comment]
Bumping this, the select should be available as a drop down, with a "side" prop like other radix components.

**Comment by jordanlesich** [comment]
Subscribed to this. Side and offset props would be great.

**Comment by haleyfrost** [comment]
I'd also love to see this. We love the select menu so far, but the positioning restrictions make it frustrating when the UI is closer to the top. It feels clunky.

**Comment by benoitgrelard** [comment]
> …but the positioning restrictions make it frustrating when the UI is closer to the top. It feels clunky.

Can you explain a bit more here? I'd love to understand what you mean given that this is how native `select` works too. Allbeit, it might happen less often as they can pop "out" of the browser on the native layer.

**Comment by benoitgrelard** [comment]
> To preface all of this, I should mention that we're currently rendering a react iframe into an ember page.

That makes sense, thanks for the explanation. Being that only one part of the page is in an iframe, that probably means you have even less available space for layout, thus more collisions.

**Comment by nosachamos** [comment]
This component has potential but until we can have the simple and standard behavior of opening below the trigger, it cannot be used in a real setting. Frequently the drop down will be placed near the top. If an item close to the bottom of the selected items is picked, the drop down behaves in a very unreasonable when opened again.

To repro just use the sample drop down, pick the last item in it. Then scroll page page so that the drop down is near the top of the viewport, and open it. Instead of seeing all options below as expected, you see just 2 or 3 and then the user has to scroll up to see the rest.

Please implement something standard - these are supposed to be primitives. Let the people change the basics if they want to :+1:

**Comment by zavjs** [comment]
I too subscribe to this request. Being able to configure the position of the dropdown underneath the Select would be extremely useful.

**Comment by third774** [comment]
I'm also running into an issue where I would like to have more control over the positioning of the content area when it opens. In my case, the select is in a control bar across the bottom of the viewport, and while the viewport is large enough to display the entire list, it only displays a few items, requiring the user to scroll to reveal the full list of options.

**Comment by benoitgrelard** [comment]
Hey all 👋,

I'd love some quick feedback from all of you who are after some more flexibility regarding positioning.

What is the primary reason you want to position options below the trigger
1. is it mostly because you do not like the fact the selected option will align with the trigger (in best case scenario)?
2. is it mostly because you do not like the fact that the user has to scroll when in edge cases (close to the top/bottom) and you'd rather show all options? (at the detriment of the selected one being far away from the trigger)
3. other?

**Please respond with 1, 2 or 3 and some more details if you wish.**

---

Basically, I'm trying to figure out whether if we provided an option to bypass the minimum height constraint (that is used to make sure the selected item isn't too far from the trigger when in edge-cases) would be enough for most of you.

ie.

**Current behaviour (which would be the default)**

https://user-images.githubusercontent.com/1539897/174257651-8acf7fbb-acf6-4581-bc56-3bc4bde8cf9b.mp4

**With option (new behaviour)**

https://user-images.githubusercontent.com/1539897/174257420-2cb4b163-c3df-452e-a6ae-87044cd10019.mp4

**Comment by gunningham** [comment]
3. I like the default behaviour but unfortunately I have to align with designs.

In my case, I need the options to always be positioned below the trigger:

![image](https://user-images.githubusercontent.com/23052507/174264639-a8696512-55d7-47f3-b83f-5caa8174a36b.png)

**Comment by Johnhersh** [comment]
#3. Other: To better align with certain formats (windows)
For me it's that our UI designers bring up designs based on common standards. And a regular HTML select looks like this on Windows:
![image](https://user-images.githubusercontent.com/20746719/174286825-959299db-baef-43bd-a7ea-ce4b8fe62b3f.png)

On a Mac, whether Chrome of Safari, it behaves like your select component here. And it seems on Android it brings up a full-screen menu, which is another beast altogether.

But the design I got from our UI department was based on the Select component as they understand it from a basic html tag, and that one behaves more like a dropdown.

**Comment by haleyfrost** [comment]
@benoitgrelard For us, #2 feels like the most obvious choice. Our biggest struggle is the collision when items are too close to the top or bottom (but, again, that's due to our react iframe inside of an ember app that limits our available real estate). That said, it would be a nice win to be able to tell it to always populate the popper below the menu but I completely understand if that functionality is a bit too much. 

Our menu will only show up to 6 options (for now) so being able to see all of them would be great, too. Thanks for collecting feedback and for your willingness to listen. We appreciate you and all your team does.

**Comment by third774** [comment]
@benoitgrelard Option `#2` would be tremendous!

**Comment by MrHappy8D** [comment]
@benoitgrelard #2 would be perfect. To give some context, a number of users are notifying us that when they click on the select menu they need to scroll up in order to see the other options — this is because the dropdown is always at the top of the screen. Have screenshot an example in safari, mac os. In this scenario the current selected option isn't important to them when opening the select menu — they're wanting to see what other options there are and swap. If there was an option to set a variant for the position that would be great!

On that note would there be a hacky way to achieve #2 with the current component?

![image](https://user-images.githubusercontent.com/29813966/177248983-dbfeeea2-5458-4f53-9b09-b583816078e2.png)

**Comment by asiraky** [comment]
#3 - My designer doesnt like the MacOS style combobox, and prefers the combox box to be below the trigger.

**Comment by pstoica** [comment]
Been following this thread for a while, along with the development of Select. I just want to chime in that I also can't adopt the Select component until it has a more menu-like layout (below the trigger; we're still using Menu as a fake Select).

- Overlaying the currently selected answer has its oddities as mentioned before. It's not always desirable compared to getting a sense of all options.
- Optimizing for filling the viewport's height feels a bit invasive (blocks more UI than intended) and too opinionated if that's the only option provided.

**Comment by BerkeKaragoz** [comment]
@benoitgrelard It is both 1 and 3 for me.
I like how select works, it looks good. But this isn't the standard behavior of the plain select.

I use the radix select for my projects and I like the way it works. But it doesn't work for every project. For example, my client has a design made and the way they designed is similar to the plain select. Which looks better for their case.

Also in the demo of the component, content shift happens when the direction arrows change states.

I don't want to give up the current behavior. However, there should be a viewport option that disables this behavior. Maybe making it as a default, and making the item offset an option, would be even better.

**Comment by benoitgrelard** [comment]
> But this isn't the standard behavior of the plain select.

It is the standard behavior on MacOS. Unfortunately there aren't any "overall" standard, so Windows will have a different way for example. So there's no easy way to make it feel natural for all your users.

**Comment by BerkeKaragoz** [comment]
> > But this isn't the standard behavior of the plain select.
> 
> It is the standard behavior on MacOS. Unfortunately there aren't any "overall" standard, so Windows will have a different way for example. So there's no easy way to make it feel natural for all your users.

I didn't know that. Maybe my "standard" wording was not correct. Even though MacOS style looks good, most GNU/Linux GUIs, Windows and maybe others which do this make select feel simpler.

**Comment by benoitgrelard** [comment]
Yeah that is the issue as ours matches the MacOS experience. But ultimately you are right, we need to provide a way to build both experiences 👍

**Comment by MatheusPires99** [comment]
It would be really nice as well if we could add something like max-height to the select options. I have a select with more then 100 options, and it takes the whole screen to display them, the UI looks very weird the way it is now.

**Comment by mrj0** [comment]
I am a huge fan of this project and have used it extensively, so I say with love: I also struggled to adopt the Select component mainly because of design decisions. Hey, I thought the design decisions would be up to me in a headless component library. But Select makes a lot of decisions, like locking the scroll (unwanted), positioning the select (I had positioning bugs that I struggled to reproduce), and even inline styles that I wasn't able to override.

Select is good -- I forked it and was able to remove the lock, absolutely position the dropdown and maintain keyboard navigation without locking scroll. It was pretty straight-forward to modify but I totally get why it's hard to support multiple behavior styles. I just think Select maybe needs to not style so much and it would be much easier. Maybe some of the behavior could fit into optional components?

**Comment by mnlfischer** [comment]
The decision where to put the dropdown should be only made by the user. You can get examples from headless ui implementation. There is no need to create custom props to define the position. Its css, not js.

**Comment by benoitgrelard** [comment]
> I am a huge fan of this project and have used it extensively, so I say with love: I also struggled to adopt the Select component mainly because of design decisions. Hey, I thought the design decisions would be up to me in a headless component library. But Select makes a lot of decisions, like locking the scroll (unwanted), positioning the select (I had positioning bugs that I struggled to reproduce), and even inline styles that I wasn't able to override.

> Select is good -- I forked it and was able to remove the lock, absolutely position the dropdown and maintain keyboard navigation without locking scroll. It was pretty straight-forward to modify but I totally get why it's hard to support multiple behavior styles. I just think Select maybe needs to not style so much and it would be much easier. Maybe some of the behavior could fit into optional components?

@mrj0 that's the plan ultimately, we'll try and provide both options.

> There is no need to create custom props to define the position. Its css, not js.

@mnlfischer unfortunately no it's not that simple. It has to be js because of things like collision handling, etc

**Comment by mnlfischer** [comment]
@benoitgrelard collision handling is something you added and is not defined in w3c standards, am I right? If yes, I think this behavior should be implemented by the user and removed from Radix to simplify your code and align it more to the html standard.

**Comment by benoitgrelard** [comment]
> @benoitgrelard collision handling is something you added and is not defined in w3c standards, am I right? If yes, I think this behavior should be implemented by the user and removed from Radix to simplify your code and align it more to the html standard.

This unfortunately wouldn't fit with the philosophy behind Radix, which isn't about just providing accessibility but also great functionality out of the box. You would be surprised by the amount of less experienced developers that use Radix for this very reason. One of our goals is specifically that ours users do not need to piece together lots of different libs to make things work. `DropdownMenu` is a good example of that.

The main issue here (which we agree with btw) is that the current option is a little too opinionated and that we should provide an alternative. This is the approach we are going to take.

**Comment by TomasSestak** [comment]
Is there a hacky way to position it under trigger??

**Comment by jhoangqm** [comment]
I am also facing the same issue on Windows, wherever I open the trigger the Select.Item is shown in the middle.

**Comment by ryparker** [comment]
> Is there a hacky way to position it under trigger??

I worked around this by sorting the options list so that the selected option is always first.

Demo from [congress.wiki](https://congress.wiki):
![CleanShot 2022-10-13 at 14 55 55](https://user-images.githubusercontent.com/17558268/195718355-c204e5a5-bebb-4038-a362-922a2c3e4da7.gif)


```ts
// Sort the selectedValue to the top of the options array.
// This is so that the menu always opens below the trigger.
options.sort(
  (a, b) =>
    (a.value === selectedValue && -1) ||
    (b.value === selectedValue && 1) ||
    0,
);
```

**Comment by benoitgrelard** [comment]
> I worked around this by sorting the options list so that the selected option is always first.

Isn't this worse for the users by not having a consistent list they can rely on? You can't build any sort of muscle memory if the order is constantly changing under you right?

**Comment by paul-sachs** [comment]
@ryparker You may be better served by using a dropdown menu with a dynamic trigger. That should give you better control of positioning . We still have a bug #1596 where if it's too large for the window, but might still be sufficient.

**Comment by surjithctly** [comment]
Trying Radix UI for the first time, had the same issue, and stumbled up on this thread. I thought I could easily position it using CSS classes, but it isn't the case. 

So I agree with other users that Radix should be un-opinionated thus provide an option for showing below the select + controllable using CSS.

**Comment by mnlfischer** [comment]
@surjithctly use headlessui, adobe aria or one of Kents or Micheals libs like downshift or reachui.

**Comment by surjithctly** [comment]
@mnlfischer using headlessui already, but radix has more components 🔥

**Comment by ColmanHumphrey** [comment]
Hey @benoitgrelard — appreciate that you're collecting feedback! For us, the fact that only a few options are visible in edge cases is the issue. Your [video above](https://github.com/radix-ui/primitives/issues/1247#issuecomment-1158624937) is exactly what we'd like to see (although I'm guessing you're all well on the way to already solving this!).

Currently, our team have found that if you need to use select components near the top a lot, this friction of having to scroll (or hover the mouse over the arrow and wait) adds up.

Thanks again for the great work!

**Comment by benoitgrelard** [comment]
🎅🎄🎁🎅🎄🎁🎅🎄🎁 Something is coming to Radix 's Select! If you know… you know!!!

https://twitter.com/benoitgrelard/status/1605205509417701376

https://user-images.githubusercontent.com/1539897/208687793-05b534a1-a302-4242-860b-8d31d0320c09.mp4

**Comment by nosachamos** [comment]
Nice!! Thanks!!

**Comment by selenecodes** [comment]
Thanks so much!

**Comment by ishaan1013** [comment]
Thank you so much!

**Comment by nosachamos** [comment]
Thank you!!!

**Comment by benoitgrelard** [comment]
For all of you for which this was a blocker, I'm happy to announce this has now been shipped! 🙂

https://www.radix-ui.com/docs/primitives/overview/releases#january-17-2023

Go update your selects!!! 🚀🚀🚀
